src/data_source/storage/pruning.rs

```diff
@@ -0,0 +1,155 @@
+use anyhow::bail;
+use async_trait::async_trait;
+use std::error::Error;
+use std::fmt::Debug;
+use std::time::Duration;
+
+#[derive(Clone, Debug)]
+pub struct PrunerCfg {
+    /// Disk space threshold (in bytes).
+    ///
+    /// If the disk usage exceeds this threshold, pruning of data starts from
+    /// the oldest data and continues until the disk usage falls below `MAX_USAGE
+    /// or until the oldest data is younger than `MINIMUM_RETENTION`
+    pruning_threshold: Option<u64>,
+
+    /// Minimum data retention period
+    ///
+    /// Data younger than this is never pruned, regardless of disk usage.
+    minimum_retention: Duration,
+
+    /// Target data retention period
+    ///
+    /// This is the ideal period for which data should be retained
+    /// data younger than this and older than `MINIMUM_RETENTION` may be pruned if disk usage exceeds the `pruning_threshold`.
+    target_retention: Duration,
+
+    /// Number of blocks to remove in a single pruning operation.
+    batch_size: u64,
+
+    /// Maximum disk usage (in basis points).
+    ///
+    /// Pruning stops once the disk usage falls below this value, even if
+    /// some data older than the `MINIMUM_RETENTION` remains. Values range
+    /// from 0 (0%) to 10000 (100%).
+    max_usage: u16,
+    /// Pruning interval
+    interval: Duration,
+}
+
+#[async_trait]
+pub trait PruneStorage: PrunerConfig {
+    type Error: Error + Debug + Send + Sync + 'static;
+
+    async fn get_disk_usage(&self) -> Result<u64, Self::Error> {
+        Ok(0)
+    }
+    async fn get_height_by_timestamp(&self, _timestamp: i64) -> Result<Option<u64>, Self::Error> {
```

Comment:
I think this function is only used as an internal helper function in the SQL pruning implementation, and never generically via this trait? If so, we could consider removing it from the trait and just making it a private method of `SqlStorage`